# Funções em Python

## Tópicos

- Introdução
- Variável local e global
- Parâmetros opcionais
- Funções como parâmetro
- Operador * (asterisco)
- Função lambda
- Recursividade

***
## Introdução

- Trecho de código associado com um nome que é executado quando é “chamado”;
- Exemplos de funções internas de Python: ``print``, ``len``, ``input``, etc;
- Podem receber argumentos ou não;
- Argumentos são informações ou valores que são enviados para as funções: números, strings, booleanos, listas, outras funções, etc;
- Parâmetros são as variáveis que são definidas dentro dos parênteses para definir as funções;

In [3]:
def minha_funcao_imprime(parametro):
    ## executa algo
    print(parametro)

argumento = "Uma mensagem qualquer."

minha_funcao_imprime(argumento)


TypeError: minha_funcao_imprime() missing 1 required positional argument: 'parametro'

- Exemplo de argumento opcional: mensagem na saída padrão da função ``input``

In [4]:
input_1 = input("Digite algo: ")
input_2 = input()
print("input_1: %s" % input_1)
print("input_2: %s" % input_2)

Digite algo: Algo
Algo2
input_1: Algo
input_2: Algo2


- Exemplo de argumento obrigatório: a função ``len`` precisa de um argumento.

In [6]:
L = list(range(10))

print(len(L))
print(L)

10
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


In [9]:
print(len())

TypeError: len() takes exactly one argument (0 given)

***

## ``args``

- Agora, vamos analisar a função ``soma`` definida da seguinte maneira:

In [15]:
def soma(a, b):
    return a + b

- Esta função retorna o valor da soma de dois números usando a palavra-chave ``return``. Na verdade, toda função em Python retorna um valor. Quando a declaração ``return`` não é especificada, a função retorna o valor ``None``;
- Esta função possui apenas dois parâmetros definidos, portanto, apenas realiza a soma de dois números. Qualquer tentativa de fazer ``soma(1,10,4)`` retornaria um erro.
- Podemos flexibilizar a definição da função para aceitar um número desconhecido de parâmetros __posicionais__. Para isto, usamos a variável ``*args``:

In [16]:
def soma(a, b, *args):
    
    soma_total = a + b
    
    print(args)
    
    for n in args:
        soma_total += n
    return soma_total

print(soma(1,2))
print(soma(1,2,3))
print(soma(45,26,369,47,88))

()
3
(3,)
6
(369, 47, 88)
575


- Note que a passagem de pelo menos dois argumentos ainda é obrigatória ``a`` e ``b``.
- Utilizar o nome ``args`` é apenas uma convenção. Podemos usar qualquer palavra, por exemplo, ``*numeros``, ``*fatores``, etc.

__Pergunta__: o que faz o asterisco?
- O asterisco extrai elementos individuais a partir de uma sequência de elementos.

In [18]:
L = list(range(1, 100, 10))
print(L)
print(*L)
print(1, 11, 21, 31, 41, 51, 61, 71, 81, 91)

[1, 11, 21, 31, 41, 51, 61, 71, 81, 91]
1 11 21 31 41 51 61 71 81 91
1 11 21 31 41 51 61 71 81 91


***

## ``kwargs``

- Enquanto que a utilização de ``*args`` permite a função utilizar um número indefinido de argumentos posicionais, ``**kwargs`` nos permite utilizar argumentos associados com uma palavra-chave.
- O asterisco único de ``args`` associa um valor e uma posição (como acontece em listas e tuplas, por exemplo);
- O asterisco duplo de ``kwargs`` associa um valor com uma palavra-chave (como acontece em dicionários);
- Exemplo:

In [20]:
def capital(**kwargs):
    for chave,valor in kwargs.items():
        print(valor, "é a capital do", chave)

capital(Brasil="Brasília", Chile="Santiago", Paraguai="Assunção", Argentina="Buenos aires")

capital()

Brasília é a capital do Brasil
Santiago é a capital do Chile
Assunção é a capital do Paraguai
Buenos aires é a capital do Argentina


***

## Parâmetros com valor padrão

- Podemos observar nos exemplos anteriores que os argumentos passados para ``args`` e ``kwargs`` não são obrigatórios. 
- Por exemplo, a função ``capital`` pode ser executada sem nenhum argumento, mas também não retorna nenhum valor.
- Por outro lado, podemos projetar uma função para utilizar um valor pré-definido como padrão na ausência de argumentos e deixando a possibilidade de alterá-lo caso necessário.
- Exemplo:

In [21]:
def gerar_lista():
    L = list(range(0,10))
    print(L)

gerar_lista()

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


In [22]:
def gerar_lista_2(inf=0, sup=10):
    L = list(range(inf,sup))
    print(L)

gerar_lista_2()
gerar_lista_2(13, 30)


[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29]


- A definição da função ``gera_lista_2`` permite utilizá-la como a função ``gera_lista``, gerando uma lista no intervalo de 0 a 9 caso não seja passado nenhum argumento. 
- Isto é possível porque foram definidos valores padrão (``inf=0`` e ``sup=10``) para parâmetros opcionais ``inf`` e ``sup``.
- Por outro lado, ``gera_lista_2`` também permite que seja criada uma lista num intervalo de números qualquer.
- Podemos usar parâmetros obrigatórios e opcionais na mesma função. Exemplo:

In [23]:
def soma(a, b, imprime = False):
    
    if imprime:
        print("A soma é {}".format(a + b))
        return ## é o mesmo que return None
    return a + b


s1 = soma(2, 5, True)

s2 = soma(2, 5)

print(s1)
print(s2)

A soma é 7
None
7


- __Importante__: parâmetros posicionais devem aparecer primeiro que parâmetros com valor padrão e opcionais:

In [ ]:
def funcao(argumentos_posicionais, argumentos_com_valor_padrao , *args, **kwargs):
    pass

- Se explicitamos o nome do parâmetro durante a chamada da função, a ordem não importa. Exemplo:

In [14]:
## Neste caso, a ordem deve ser preservada

def imprime_repeticao(numero, repeticoes=10):
    for i in range(repeticoes):
        print(numero)

## repetir 10 vezes o número 1 

imprime_repeticao(1,10)
print("#################################")
imprime_repeticao(10,1)
print("#################################")
imprime_repeticao(repeticoes=10,numero=1)

1
1
1
1
1
1
1
1
1
1
#################################
10
#################################
1
1
1
1
1
1
1
1
1
1


- Ao passar o nome do parâmetro na chamada da função, todos os parâmetros tem que ser nomeados também.

In [15]:
# erro

imprime_repeticao(10,numero=1)

TypeError: imprime_repeticao() got multiple values for argument 'numero'

In [16]:
# erro também
imprime_repeticao(numero=1, 10)

SyntaxError: positional argument follows keyword argument (<ipython-input-16-4b828b4bb02e>, line 2)

- No entanto, devemos ter cuidado quando temos parâmetros obrigatórios e opcionais:
 - Podemos passar os parâmetros obrigatórios sem usar os seus nomes, mas devemos respeitar a ordem usada na definição da função; e nomeando os parâmetros opcionais:

In [ ]:
imprime_repeticao(1, repeticoes=5)

## Funções como parâmetros

- Funções também podem ser usadas como parâmetros para outras funções;
- Exemplo:

In [26]:

def modifica_string(string1, string2, operacao):
    print("Resultado da modificação: ", operacao(string1, string2))
    
def concatena_string(string1, string2):
    return string1 + string2

def concatena_maiuscula_string(string1, string2):
    return string1.upper() + string2.upper()

def concatena_lista_string(string1, string2):
    V = string1 + string2
    
    return list(V)

modifica_string("Tudo vai ", "passar!", concatena_string)
modifica_string("Tudo vai ", "passar!", concatena_maiuscula_string)
modifica_string("Tudo vai ", "passar!", concatena_lista_string)

NameError: name 'V' is not defined

## Variável local e global

- Variáveis possuem escopos, ou seja, áreas do programa onde são visíveis.
- Uma variável que existe apenas dentro de uma função só é visível internamente na função. Portanto, não podemos acessá-la fora do escopo da função onde ela é definida, ou seja, é uma __variável local__.
- Uma __variável global__ é definida fora do escopo de funções. É visível em todo módulo e por outros módulos que importarem o módulo onde está definida.
- Boa prática 1: use letras maiúsculas para representar variáveis globais;
- Boa prática 2: preferencialmente usar a palavra-chave ``global`` apenas para alterar variáveis associadas com a configuração do módulo ou programa.


### Exemplos de variáveis locais e globais
#### Variável global

In [28]:
x = "global_2020"

def func():
    print("x dentro da função:", x)


func()
print("x fora da função:", x)

x dentro da função: global_2020
x fora da função: global_2020


#### Modificando a variável global dentro de uma função

In [31]:
x = "global_2021"

def func():
    x = x * 2
    print(x)

func()

UnboundLocalError: local variable 'x' referenced before assignment

- Neste último caso, temos um erro porque o interpretador entende a variável ``x`` como variável local, mas não encontra nenhuma inicialização para ``x`` dentro da função ``func``.
- Para realizar essa alteração, utilizamos a palavra-chave ``global``:

In [35]:
x = "global"

def func():
    x = "não é mais global"
    print(x)

func()
print(x)

não é mais global
global


#### Variável local

- Vamos tentar acessar uma variável local fora do escopo da função onde foi definida:

In [36]:
def func():
    y = "local"


func()
print(y)

NameError: name 'y' is not defined

- Como esperado, obtemos um erro porque a variável ``y`` não está definida no escopo em que está sendo referenciada.

#### Variáveis locais e globais com o mesmo nome

In [39]:
x = 5

def func():
    x = 10
    print("local x:", x)


func()
print("global x:", x)

local x: 10
global x: 5


#### Variável ``nonlocal``

- Uma variável ``nonlocal`` são usadas em funções dentro de funções que não estão definidas no escopo local.
- Parece confuso, mas isto significa que a variável não está no escopo local e nem no escopo global.
- Exemplo:

In [21]:
## escopo global
### x não está aqui

def func_mais_externa():
    x = "local"

    def func_mais_interna():
        nonlocal x
        x = "nonlocal"
        print("mais interna:", x)

    func_mais_interna()
    print("mais externa:", x)

func_mais_externa()

mais interna: nonlocal
mais externa: nonlocal


## Função ``lambda``

- Função ``lambda`` é uma função simples para realizar operações igualmente simples.
- Uma função ``lambda`` é definida para ser usada uma única vez para resolver algo simples em determinado trecho do código;
- Sua utilização é um pouco controversa, porque insere definições de funções no "meio do código";
- São usadas para fazer pequenos cálculos ou processamentos de entradas, por exemplo.
- Relembre a função ``concatena_maiuscula_string`` das seções anteriores:

In [20]:
def concatena_maiuscula_string(string1, string2):
    return string1.upper() + string2.upper()

- Uma função ``lambda`` correspodente seria:

In [40]:
concatena_maiuscula_string_lambda = lambda string1, string2: string1.upper() + string2.upper()

s = concatena_maiuscula_string_lambda("Bom dia, ", "equipe!")

print(s)

BOM DIA, EQUIPE!


- Vamos criar uma função anônima dentro de uma outra função que seja multiplicadora de números ``minha_funcao``;
- ``minha_funcao`` recebe um argumento e o argumento será multiplicado por um número desconhecido;

In [42]:
def minha_funcao(n):
    return lambda a : a * n

In [43]:
## define n como 2
multiplica_por_2 = minha_funcao(2)

## define n como 5
multiplica_por_5 = minha_funcao(5)

## define n como 10
multiplica_por_10 = minha_funcao(10)

print(multiplica_por_2(10))
print(multiplica_por_5(3))
print(multiplica_por_10(3))

20
15
30


## Recursividade

- Recursão é um método para resolver problemas quebrando o problema original em instâncias menores do mesmo problema;
- A solução dos sub-problemas são combinadas para se obter a solução do problema original.
- Uma função recursiva é uma função que chama a si mesma para realizar alguma operação. A cada chamada, a função resolve um sub-problema;
- Pode ser muito eficiente para resolver problemas matemáticos;
- Requer cautela para não criar uma função que execute indefinidamente realizando sucessivas chamadas a si mesmo sem um ponto de parada adequado.
- Portanto, deve possuir uma regra ou condição de parada da recursão;
- A memória utilizada durante cada chamada é um ponto a ser considerado.

### Criando uma função recursiva

- Existem duas etapas importantes para criação de uma função que resolve um problema recursivamente:

    - __Definir um "critério de parada" ou caso base__: o menor problema que pode ser resolvido sem realizar mais nenhuma sub-divisão do problema atual. É um caso onde a função apenas retorna o resultado e não é necessário mais nenhuma redução do problema.
    - __Definir um "passo de redução" ou caso recursivo__: a parte do código onde a função chama a si mesma para tentar resolver um problema menor.

### Caso clássico: calcular fatorial de um número

- Vamos implementar uma função que calcula recursivamente o fatorial de um número.
- Relembrando o cálculo do fatorial de alguns números:

$ 5! = 5 \times 4 \times 3 \times 2 \times 1 = 5 \times 4! $ <br>

$ 4! = 4 \times 3 \times 2 \times 1 = 4 \times 3! $ <br>

E assim por diante.



- Isto significa que podemos calcular o fatorial de um número multiplicando este número pelo fatorial do número anterior 💡

$ N! = N \times (N - 1)! $ <br>

- Como falado anteriormente, para criar uma função recursiva precisamos  definir o __caso recursivo__ e o __caso base__.

#### Definindo o caso recursivo

- Precisamos definir um passo onde vamos tentar encontrar a solução em instâncias mais simples do nosso problema.
- No caso do fatorial de um número ``n``, este caso seria a multiplicação de ``n`` pelo fatorial de ``n - 1`` (redução).

In [44]:
def fatorial_recursivo(n):
    # Caso recursivo
    ###print(n)
    return n * fatorial_recursivo(n-1)

fatorial_recursivo(3)

3
2
1
0
-1
-2
-3
-4
-5
-6
-7
-8
-9
-10
-11
-12
-13
-14
-15
-16
-17
-18
-19
-20
-21
-22
-23
-24
-25
-26
-27
-28
-29
-30
-31
-32
-33
-34
-35
-36
-37
-38
-39
-40
-41
-42
-43
-44
-45
-46
-47
-48
-49
-50
-51
-52
-53
-54
-55
-56
-57
-58
-59
-60
-61
-62
-63
-64
-65
-66
-67
-68
-69
-70
-71
-72
-73
-74
-75
-76
-77
-78
-79
-80
-81
-82
-83
-84
-85
-86
-87
-88
-89
-90
-91
-92
-93
-94
-95
-96
-97
-98
-99
-100
-101
-102
-103
-104
-105
-106
-107
-108
-109
-110
-111
-112
-113
-114
-115
-116
-117
-118
-119
-120
-121
-122
-123
-124
-125
-126
-127
-128
-129
-130
-131
-132
-133
-134
-135
-136
-137
-138
-139
-140
-141
-142
-143
-144
-145
-146
-147
-148
-149
-150
-151
-152
-153
-154
-155
-156
-157
-158
-159
-160
-161
-162
-163
-164
-165
-166
-167
-168
-169
-170
-171
-172
-173
-174
-175
-176
-177
-178
-179
-180
-181
-182
-183
-184
-185
-186
-187
-188
-189
-190
-191
-192
-193
-194
-195
-196
-197
-198
-199
-200
-201
-202
-203
-204
-205
-206
-207
-208
-209
-210
-211
-212
-213
-214
-215
-216
-217
-218
-219
-220


-1559
-1560
-1561
-1562
-1563
-1564
-1565
-1566
-1567
-1568
-1569
-1570
-1571
-1572
-1573
-1574
-1575
-1576
-1577
-1578
-1579
-1580
-1581
-1582
-1583
-1584
-1585
-1586
-1587
-1588
-1589
-1590
-1591
-1592
-1593
-1594
-1595
-1596
-1597
-1598
-1599
-1600
-1601
-1602
-1603
-1604
-1605
-1606
-1607
-1608
-1609
-1610
-1611
-1612
-1613
-1614
-1615
-1616
-1617
-1618
-1619
-1620
-1621
-1622
-1623
-1624
-1625
-1626
-1627
-1628
-1629
-1630
-1631
-1632
-1633
-1634
-1635
-1636
-1637
-1638
-1639
-1640
-1641
-1642
-1643
-1644
-1645
-1646
-1647
-1648
-1649
-1650
-1651
-1652
-1653
-1654
-1655
-1656
-1657
-1658
-1659
-1660
-1661
-1662
-1663
-1664
-1665
-1666
-1667
-1668
-1669
-1670
-1671
-1672
-1673
-1674
-1675
-1676
-1677
-1678
-1679
-1680
-1681
-1682
-1683
-1684
-1685
-1686
-1687
-1688
-1689
-1690
-1691
-1692
-1693
-1694
-1695
-1696
-1697
-1698
-1699
-1700
-1701
-1702
-1703
-1704
-1705
-1706
-1707
-1708
-1709
-1710
-1711
-1712
-1713
-1714
-1715
-1716
-1717
-1718
-1719
-1720
-1721
-1722
-1723
-1724
-172

ERROR:root:Internal Python error in the inspect module.
Below is the traceback from this internal error.



Traceback (most recent call last):
  File "C:\Users\vitor\miniconda3\lib\site-packages\IPython\core\interactiveshell.py", line 3417, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-44-fd06ec041d8e>", line 6, in <module>
    fatorial_recursivo(3)
  File "<ipython-input-44-fd06ec041d8e>", line 4, in fatorial_recursivo
    return n * fatorial_recursivo(n-1)
  File "<ipython-input-44-fd06ec041d8e>", line 4, in fatorial_recursivo
    return n * fatorial_recursivo(n-1)
  File "<ipython-input-44-fd06ec041d8e>", line 4, in fatorial_recursivo
    return n * fatorial_recursivo(n-1)
  [Previous line repeated 2954 more times]
  File "<ipython-input-44-fd06ec041d8e>", line 3, in fatorial_recursivo
    print(n)
  File "C:\Users\vitor\miniconda3\lib\site-packages\ipykernel\iostream.py", line 404, in write
    self.pub_thread.schedule(lambda : self._buffer.write(string))
  File "C:\Users\vitor\miniconda3\lib\site-packages\ipykernel\iostream.py", line 202, in sc

TypeError: object of type 'NoneType' has no len()

- Experimente descomentar a linha 3 e 6 do código anterior e observe a saída.

#### Definindo o caso base

- A definição anterior da função ``fatorial_recursivo`` fará com que o interpretador execute a função indefinidamente enquanto houver recursos computacionais disponíveis.
- Para evitar isso, precisamos definir a condição de parada ou dar a solução do menor problema resolvido pela função.
- No caso do cáculo de fatorial, a problema mais simples a ser resolvido é o fatorial de 0 (zero).
- Vamos atualizar a definição da função ``fatorial_recursivo`` adicionando o caso base:

In [26]:
def fatorial_recursivo(n):
    # Caso base   
    if n == 0:
        return 1
    # Caso recursivo
    else:
        return n * fatorial_recursivo(n-1)

- Este é um exemplo clássico de algoritmo recursivo: se você tem um problema simples, resolva. Senão, divida-o em sub-problemas e aplique a mesma estratégia de resolução anterior.

##### Vamos pensar um mais pouco sobre recursão

1 - Considere a função recursiva de fatorial abaixo:

In [27]:
def fatorial_recursivo(n):
    if n < 0:
        print("Sem solução.")
    elif n == 0:
        return 1
    else:
        return n * fatorial_recursivo(n-1)

- Considere que executamos ``fatorial_recursivo(4)``. Qual a ordem de execução das chamadas de funções abaixo? <br>
[ ] ``fatorial_recursivo(3)`` <br>
[ ] ``fatorial_recursivo(0)`` <br>
[ ] ``fatorial_recursivo(2)`` <br>
[ ] ``fatorial_recursivo(4)`` <br>
[ ] ``fatorial_recursivo(1)`` <br>

### Exercício: os números de Fibonacci

- Os números de Fibonacci ($F_{n}$) formam uma sequência onde cada número é a soma dos dois números que o precedem: 0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, ... Ou seja:

$ F_n = F_{n−1} + F_{n−2} $ e $F_0 = 0$, $F_1 = 1$.

- Vamos usar a definição dos números de Fibonacci para escrever nossa função:

In [ ]:
def fib(n):

    if n == 0: ## caso base
        return ... 
    elif n == 1: ## caso base
        return ...
    else:      
        return ... ## façam o caso recursivo!